Data/Primitive/PrimArray.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+{-# OPTIONS_GHC -Wall #-}
+
+module Data.Primitive.PrimArray
+  ( -- * Types
+    PrimArray(..)
+  , MutablePrimArray(..)
+    -- * Allocation
+  , newPrimArray
+    -- * Element Access
+  , readPrimArray
+  , writePrimArray
+  , indexPrimArray
+    -- * Freezing and Thawing
+  , unsafeFreezePrimArray
+  , unsafeThawPrimArray
+    -- * Block Operations
+  , copyPrimArray
+  , copyMutablePrimArray
+  , copyPrimArrayToPtr
+  , copyMutablePrimArrayToPtr
+  , setPrimArray
+    -- * Information
+  , sameMutablePrimArray
+  , getSizeofMutablePrimArray
+  , sizeofPrimArray
+  ) where
+
+import GHC.Prim
+import GHC.Exts (isTrue#,IsList(..))
+import GHC.Int
+import GHC.Ptr
+import Data.Primitive
+import Control.Monad.Primitive
+import Control.Monad.ST
+import Data.Semigroup (Semigroup(..))
+import qualified Data.Semigroup as SG
+import qualified Data.List as L
+import qualified Data.Primitive.ByteArray as PB
+import qualified Data.Primitive.Types as PT
+
+-- | Primitive arrays
+data PrimArray a = PrimArray ByteArray#
+
+-- | Mutable primitive arrays associated with a primitive state token
+data MutablePrimArray s a = MutablePrimArray (MutableByteArray# s)
+
+instance (Eq a, Prim a) => Eq (PrimArray a) where
+  a1 == a2 = sizeofPrimArray a1 == sizeofPrimArray a2 && loop (sizeofPrimArray a1 - 1)
+    where 
+    loop !i | i < 0 = True
+            | otherwise = indexPrimArray a1 i == indexPrimArray a2 i && loop (i-1)
```

Comment:
I think it's reasonable to require that the underlying `==` be an equivalence relation (here and for the other types). Reflexivity then allows us to do a quick pointer equality check with `sameWhateverArray`.

Reply:
I think that's reasonable. Let me make sure I understand you correctly. You're saying that I would keep the `Prim` constraint and the currently implementation but would just a `sameByteArray` check that happens first to see if we can skip the element-by-element check. Is that correct?

Reply:
That's my suggestion, yes.